h2o-algos/src/main/java/hex/tree/SharedTreeModelGBMDRF.java

```diff
@@ -0,0 +1,159 @@
+package hex.tree;
+
+import hex.genmodel.algos.tree.SharedTreeNode;
+import hex.genmodel.algos.tree.SharedTreeSubgraph;
+import hex.genmodel.algos.tree.*;
+import water.*;
+import water.fvec.Chunk;
+import water.fvec.Frame;
+import water.fvec.NewChunk;
+import water.fvec.Vec;
+import water.util.*;
+
+import java.util.ArrayList;
+import java.util.List;
+import hex.Model;
+import water.Key;
+
+public abstract class SharedTreeModelGBMDRF<
+        M extends SharedTreeModel<M, P, O>,
+        P extends SharedTreeModel.SharedTreeParameters,
+        O extends SharedTreeModel.SharedTreeOutput
+        > extends SharedTreeModel<M, P, O> implements Model.Contributions{
+    
+    public SharedTreeModelGBMDRF(Key<M> selfKey, P parms, O output) {
+        super(selfKey, parms, output);
+    }
+    
+  @Override
+  public Frame scoreContributions(Frame frame, Key<Frame> destination_key) {
+    if (_output.nclasses() > 2) {
+      throw new UnsupportedOperationException(
+              "Calculating contributions is currently not supported for multinomial models.");
+    }
+
+    Frame adaptFrm = new Frame(frame);
+    adaptTestForTrain(adaptFrm, true, false);
+    // remove non-feature columns
+    adaptFrm.remove(_parms._response_column);
+    adaptFrm.remove(_parms._fold_column);
+    adaptFrm.remove(_parms._weights_column);
+    adaptFrm.remove(_parms._offset_column);
+
+    final String[] outputNames = ArrayUtils.append(adaptFrm.names(), "BiasTerm");
+    
+    if (_parms.algoName().equals("DRF")) {
```

Comment:
the if is unnecessary and can be replaced with polymorphism, then also

    .doAll(outputNames.length, Vec.T_NUM, adaptFrm)
     .outputFrame(destination_key, outputNames, null);


will not need to duplicated. you can do

    return getScoreContributionsTask(args).doAll(...).outputFrame(...)

and then

    abstract getScoreContributionsTask(..)

and keep the common part of ScoreContributionsTask in this class and in DRF and GBM have DRFModel.ScoreContributionsTask extends ScoreContributionsTask

Reply:
Agree. Just pushed this: https://github.com/h2oai/h2o-3/pull/3628/commits/ddb6a071d58f62d0f64eee8b6174d34874eacc1f